protocol/wavefront/wavefrontlistener.go

```diff
@@ -0,0 +1,255 @@
+	return []*datapoint.Datapoint{
+		sfxclient.Cumulative("invalid_datapoints", nil, atomic.LoadInt64(&listener.stats.invalidDatapoints)),
+		sfxclient.Cumulative("total_connections", nil, atomic.LoadInt64(&listener.stats.totalConnections)),
+		sfxclient.Gauge("active_connections", nil, atomic.LoadInt64(&listener.stats.activeConnections)),
+		sfxclient.Cumulative("idle_timeouts", nil, atomic.LoadInt64(&listener.stats.idleTimeouts)),
+		sfxclient.Cumulative("retry_listen_errors", nil, atomic.LoadInt64(&listener.stats.retriedListenErrors)),
+	}
+}
+
+// Close the exposed wavefront port
+func (listener *Listener) Close() error {
+	var err error
+	if listener.psocket != nil {
+		err = listener.psocket.Close()
+	}
+
+	listener.wg.Wait()
+	return err
+}
+
+type wavefrontListenConn interface {
+	io.Reader
+	Close() error
+	SetDeadline(t time.Time) error
+	RemoteAddr() net.Addr
+}
+
+// TODO need to run this and prometheus through some benchmarks for allocations
+func extractCollectdDimensions(doit bool, metricName string) (string, map[string]string) {
+	dimensions := make(map[string]string)
+	if !doit {
+		return metricName, dimensions
+	}
+	index := strings.Index(metricName, "..")
+	var toAddDims map[string]string
+	for {
+		metricName, toAddDims = collectd.GetDimensionsFromName(&metricName)
+		if len(toAddDims) == 0 {
+			// we may have put two dots next to each other extracting this stuff, see if there were two beforehand and if not, replace them if they occur with a single dot
+			// TODO could do better here if wavefront users prefixed all metrics with collectd, we could know how they're constructed and could make them be almost as good as coming from collectd
+			// could even be crazy and put the types.db file in here and provide accurate metric types... overkill?
+			if index == -1 {
+				metricName = strings.Replace(metricName, "..", ".", -1)
+			}
+
+			return metricName, dimensions
+		}
+		for k, v := range toAddDims {
+			dimensions[k] = v
+		}
+	}
+}
+
+// https://docs.wavefront.com/wavefront_data_format.html
+func (listener *Listener) fromWavefrontDatapoint(line string) *datapoint.Datapoint {
+	// TODO rewrite without Split using Index and slices as they don't allocate
+	pieces := strings.Split(line, " ")
+	if len(pieces) >= 3 {
+		metricName, dimensions := extractCollectdDimensions(listener.extractCollectdDimensions, pieces[0])
+
+		valueString := pieces[1]
+
+		var value datapoint.Value
+		if i, err := strconv.ParseInt(valueString, 10, 64); err == nil {
+			value = datapoint.NewIntValue(i)
+		} else if f, err := strconv.ParseFloat(valueString, 64); err == nil {
+			value = datapoint.NewFloatValue(f)
+		} else {
+			return nil
+		}
+		var timestamp time.Time
+		epoch, err := strconv.ParseInt(pieces[2], 10, 64)
+		if err != nil {
+			// probably not the timestamp, probably a dimension
+			pieces = pieces[2:]
+			timestamp = time.Now()
+		} else {
+			timestamp = time.Unix(epoch, 0)
+			pieces = pieces[3:]
+		}
+		for _, p := range pieces {
```

Comment:
Do we want to do anything specific for `source=`? Is it allowed to have a pointTag that's also `source=something`?

Reply:
source would just end up in our system as a dimension. in this case it could be overwritten by a pointTag yeah. I feel like that's user error and acceptable losses. According to wavefront's own write_collectd, you don't need a "source=" they use "host=" in it. So /shrug.